src/features/courses/course-controller.ts

```diff
@@ -3130,26 +3138,26 @@ class CourseController {
     }
 
     async enroll(options: CreateGradesForUserEnrollmentOptions): Promise<StudentEnrollment> {
-        const fetchedEnrollment = await StudentEnrollment.findOne({
-            where: {
-                courseId: options.courseId,
-                userId: options.userId,
-            }
-        });
-        if (!_.isNil(fetchedEnrollment)) {
-            if (!_.isNil(fetchedEnrollment.dropDate)) {
-                if (options.reEnrollIfDropped ?? false) {
-                    fetchedEnrollment.dropDate = null;
-                    await fetchedEnrollment.save();
-                } else {
-                    throw new IllegalArgumentException('Cannot enroll with enrollment link if you have dropped the course');
+        return await useDatabaseTransaction(async () => {
+            const fetchedEnrollment = await StudentEnrollment.findOne({
+                where: {
+                    courseId: options.courseId,
+                    userId: options.userId,
+                }
+            });
+
+            if (!_.isNil(fetchedEnrollment)) {
+                if (!_.isNil(fetchedEnrollment.dropDate)) {
+                    if (options.reEnrollIfDropped ?? false) {
```

Comment:
nit: Could probably condense the first two into one check, and convert the `if x else throw` into a `if !x throw`.